src/transaction.js

```diff
@@ -19,238 +16,121 @@ import { Keypair } from './keypair';
+export class Transaction extends TransactionBase {
   constructor(envelope, networkPassphrase) {
     if (typeof envelope === 'string') {
       const buffer = Buffer.from(envelope, 'base64');
       envelope = xdr.TransactionEnvelope.fromXDR(buffer);
     }
 
-    // Deprecation warning. TODO: remove optionality with next major release.
-    if (networkPassphrase === null || networkPassphrase === undefined) {
-      console.warn(
-        'Global `Network.current()` is deprecated. Please pass explicit argument instead, e.g. `new Transaction(envelope, Networks.PUBLIC)` (see https://git.io/fj9fG for more info).'
-      );
-    } else if (typeof networkPassphrase !== 'string') {
+    const envelopeType = envelope.switch();
+    if (
+      !(
+        envelopeType === xdr.EnvelopeType.envelopeTypeTxV0() ||
+        envelopeType === xdr.EnvelopeType.envelopeTypeTx()
+      )
+    ) {
       throw new Error(
-        `Invalid passphrase provided to Transaction: expected a string but got a ${typeof networkPassphrase}`
+        `Invalid TransactionEnvelope: expected an envelopeTypeTxV0 or envelopeTypeTx but received an ${envelopeType.name}.`
       );
     }
-    this._networkPassphrase = networkPassphrase;
 
     const txEnvelope = envelope.value();
-    let sourceAccount;
-    this._envelopeType = envelope.switch();
-    switch (this._envelopeType) {
-      case xdr.EnvelopeType.envelopeTypeTxV0():
-        sourceAccount = txEnvelope.tx().sourceAccountEd25519();
-        break;
-      case xdr.EnvelopeType.envelopeTypeTx():
-        sourceAccount = txEnvelope
-          .tx()
-          .sourceAccount()
-          .ed25519();
-        break;
-      case xdr.EnvelopeType.envelopeTypeTxFeeBump():
-        sourceAccount = txEnvelope
-          .tx()
-          .feeSource()
-          .ed25519();
-        break;
-      default:
-        throw new Error(`Invalid EnvelopeType: ${this._envelopeType.name}.`);
-    }
-
-    // since this transaction is immutable, save the tx
-    this.tx = txEnvelope.tx();
-    let tx = this.tx;
+    const tx = txEnvelope.tx();
+    const fee = tx.fee().toString();
+    const signatures = (txEnvelope.signatures() || []).slice();
 
-    this.source = StrKey.encodeEd25519PublicKey(sourceAccount);
-    this.fee = tx.fee().toString();
-
-    // if tx is FeeBump, override tx with innerTx, which we use to expose data
-    // about the innerTx like operations and memo.
-    if (this.isFeeBump()) {
-      const innerTxEnvelope = tx.innerTx().value();
-      tx = innerTxEnvelope.tx();
-
-      // make inner transaction the source account and add field feeSource
-      this.feeSource = StrKey.encodeEd25519PublicKey(sourceAccount);
-
-      this.source = StrKey.encodeEd25519PublicKey(tx.sourceAccount().ed25519());
-      this.innerSignatures = map(innerTxEnvelope.signatures() || [], (s) => s);
-      this.innerFee = tx.fee().toString();
-    }
+    super(tx, signatures, fee, networkPassphrase);
 
+    this._envelopeType = envelopeType;
```

Comment:
apparently ecmascript supports private class variables but I'm not sure how widely supported the feature is in browsers:

https://stackoverflow.com/a/52237988

Reply:
This is still a proposal. We could use babel, but at the end of the day it will transpile down to today's JS. I'll add a ticket to this library to start exploring this.